src/main/scala/is/hail/linalg/BlockMatrix.scala

```diff
@@ -1147,22 +1157,30 @@ class WriteBlocksRDD(path: String,
       var i = 0
       while (it.hasNext && i < nRowsInBlock) {
         val rv = it.next()
-        fullRow.set(rv)
-        ec.set(1, row)
-        val gs = fullRow.getAs[IndexedSeq[Any]](entriesIndex)
+        val region = rv.region
+
+        val entryArrayOffset = rvRowType.loadField(rv, entryArrayIdx)
+
         var blockCol = 0
-        var sampleIndex = 0
+        var colIdx = 0
         while (blockCol < gp.nBlockCols) {
           val n = gp.blockColNCols(blockCol)
           var j = 0
           while (j < n) {
-            ec.set(2, sampleAnnotationsBc.value(sampleIndex))
-            ec.set(3, gs(sampleIndex))
-            f() match {
-              case null => fatal(s"Entry expr must be non-missing. Found missing value for col $j and row $row}")
-              case t => data(j) = t.toDouble
+            if (entryArrayType.isElementDefined(region, entryArrayOffset, colIdx)) {
+              val entryOffset = entryArrayType.loadElement(region, entryArrayOffset, colIdx)
+              if (entryType.isFieldDefined(region, entryOffset, fieldIdx)) {
+                val fieldOffset = entryType.loadField(region, entryOffset, fieldIdx)
+                data(j) = loadAsDouble(region, fieldOffset)
+              } else {
+                val rowIdx = blockRow * blockSize + i
+                fatal(s"Cannot create BlockMatrix: missing value at row $rowIdx and col $colIdx")
```

Comment:
might be nice to add a `fill_na` option which fills missing values with some parameterized thing like 0 or nan. Out of scope here, though